From ab60ab5fec6f51d1b065ef904021c8ad22d3f610 Mon Sep 17 00:00:00 2001
From: Stefan Lange <s.lange@gateware.de>
Date: Tue, 29 Nov 2016 10:47:33 +0100
Subject: [PATCH] TQMT104x: automatic read MAC address from EEPROM

Automatically get MAC address from on-module I2C EEPROM
and set environment ("ethxaddr") accordingly.
Only one MAC address ist stored in EEPROM.
The other MAC addresses are generated by incrementing it +1
for each of the 5 MACs present on the T104x.

Signed-off-by: Stefan Lange <s.lange@gateware.de>
---
 board/tqc/tqmt104x/eth.c | 59 ++++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 59 insertions(+)

diff --git a/board/tqc/tqmt104x/eth.c b/board/tqc/tqmt104x/eth.c
index a50e8ec..eef40ff 100644
--- a/board/tqc/tqmt104x/eth.c
+++ b/board/tqc/tqmt104x/eth.c
@@ -14,9 +14,65 @@
 #include <asm/fsl_dtsec.h>
 #include <phy.h>
 #include <vsc9953.h>
+#include <i2c.h>
+#include <hwconfig.h>
+
+#define CONFIG_SYS_MODULE_EEPROM         	0x57
+#define CONFIG_SYS_MODULE_EEPROM_ADLEN          0x2
+#define CONFIG_SYS_MODULE_EEPROM_MAC_AD_OFFSET  0x0
 
 static int eth_phy_qsgmii_reprogram = 0;
 
+static int _tq_get_macaddress_from_eeprom(void) {
+
+	int err;
+	int i,j;
+	uint8_t ethAddrBuf[6];
+	char ethAddrStr[18]; /* xx:xx:xx:xx:xx:xx\0 */
+	const char *varname;
+	unsigned int oldI2cBus;
+
+	/* Read MAC address from EEPROM */
+	oldI2cBus = i2c_get_bus_num();
+	i2c_set_bus_num(0);
+	err = i2c_read(CONFIG_SYS_MODULE_EEPROM, CONFIG_SYS_MODULE_EEPROM_MAC_AD_OFFSET, \
+			CONFIG_SYS_MODULE_EEPROM_ADLEN, ethAddrBuf, sizeof(ethAddrBuf));
+	i2c_set_bus_num(oldI2cBus);
+
+	if(err) {
+		printf("Could not read MAC addresses: %d\n", err);
+		return err;
+	} else {
+		for(i = 0; i < 5; i++) {
+			if(i == 0) {
+				varname = "ethaddr";
+			} else if(i == 1) {
+				varname = "eth1addr";
+			} else if(i == 2) {
+				varname = "eth2addr";
+			} else if(i == 3) {
+				varname = "eth3addr";
+			} else if(i == 4) {
+				varname = "eth4addr";
+			}
+			sprintf(
+				ethAddrStr, "%02x:%02x:%02x:%02x:%02x:%02x",
+				ethAddrBuf[0], ethAddrBuf[1], ethAddrBuf[2],
+				ethAddrBuf[3], ethAddrBuf[4], ethAddrBuf[5]
+			);
+			setenv(varname, ethAddrStr);
+
+			/* Increment MAC address by 1 with overflow. */
+			j = sizeof(ethAddrBuf);
+			do {
+				j--;
+				ethAddrBuf[j]++;
+			} while(ethAddrBuf[j] == 0 && j > 0);
+		}
+	}
+	return 0;
+}
+
 /*
  * Set RGMII delay and other values in ethernet phys
  * located on the STKTxxxx starterkit
@@ -215,6 +271,9 @@ int board_eth_init(bd_t *bis)
         }
 #endif
 
+        /* get MAC addresses from I2C EEPROM */
+        _tq_get_macaddress_from_eeprom();
+
 	cpu_eth_init(bis);
 #endif
 
-- 
1.9.1

